### DISCUSSION TITLE
Vendoring dependencies in manifest mode

### DISCUSSION THREAD

**Discussion by sir-earl** [microsoft/vcpkg#30290] [Q&A] [answered]
Most package managers have a notion of vendoring a projects dependencies and allowing them to be committed to source control, so that anyone who clones that library (including CI/CD tooling) can clone and build the project without having to re-build of fetch the dependencies. I'm a little confused about a couple of things:

1. In manifest mode, vcpkg will install dependencies into the project (i.e. `vcpkg_installed`), but I couldn't find any mention of whether this folder should or even can be checked into source control, and if so, whether vcpkg will treat this like a vendored set of dependencies, and avoid having to re-fetch and build them from scratch.
2. There appears to be another feature called 'Binary Caching' that looks to be similar to vendoring, in the sense that it allows a snapshot of dependencies to be re-used across different environments, but what's the point in this, when we already have the dependencies installed locally to the project in the `vcpkg_installed` folder? Is this feature just for when you want to share a set of binaries between multiple projects, or when not using manifest mode?

**Comment by autoantwort** [comment]
1. Don't check `vcpkg_installed` in. If you want to avoid re-fetching and build from scratch, use binary caching. 
2. Maybe you have multiple projects using the same dependency locally, then you don't have to rebuild the dependencies.

**Reply by sir-earl** [reply]
Thanks for that - so when using binary caching, what's the recommended way to share this cache with a team - can this be checked into source control, but outside the project itself?

Can't help feel like it's going to result in a duplicate copy of all the binaries my team needs for their project, and that `vcpkg_installed` folder looks a lot like a vendored set of dependencies. What was the rational behind not allowing the project local packages to be used for this purpose instead?

**Reply by autoantwort** [reply]
Select one of the [providers](https://learn.microsoft.com/en-us/vcpkg/users/binarycaching) that fits your use case. And I suggest to not check in the configuration, because if you change it you have to rewrite your git history. 

`vcpkg_installed` is like `node_modules` for npm, I would not say that `node_modules` are vendored dependencies if you don't check them in into source control.

**Reply by sir-earl** [reply]
Sure, but the difference with npm is that if you do want to do vendoring, then you would check the `node_modules` folder in to source control. Here you're saying that instead of that, we need to set up an alternative, essentially duplicate copy of these dependencies, instead of just reusing what's already sitting with the project. Just trying to understand why it was designed this way, as it was presumably done like this for a reason.

Don't get me wrong, I'm enjoying vcpkg so far, and it's miles ahead of manually juggling dependencies without a package manager. I'm mostly wondering if there's a paradigm shift that I'm missing that will make things click into place conceptually.

**Reply by autoantwort** [reply]
Because you don't have a stable abi in c++, you don't have this problem in javascript

**Reply by autoantwort** [reply, marked as answer]
So if you simply check vcpkg_installed in and use it with an different compiler on another computer with a different compiler version, things could break (or crash at runtime)

**Reply by Neumann-A** [reply]
@sir-earl you want to use classic mode. The you can check in the installed tree. But don't complain if stuff breaks on your side due to using vcpkg in a way it is not supposed to be used in ;) (Why not manifest mode? Compiler changes will invalidate the installed tree and force a rebuild; classic modes never rebuilds)

**Reply by sir-earl** [reply]
@autoantwort Ah yes of course. This makes a ton of sense, thanks!
